logictest: setup-flow-async span is used after finish

pkg/ccl/logictestccl/tests/fakedist/fakedist_test.TestCCLLogic_new_schema_changer [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7784384?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7784384?buildTab=artifacts#/) on master @ [db71e045cc3f2937bf20170d254dd7e54a620ffb](https://github.com/cockroachdb/cockroach/commits/db71e045cc3f2937bf20170d254dd7e54a620ffb):


```
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:181
*   | github.com/cockroachdb/cockroach/pkg/util/tracing.WithParent
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/span_options.go:204
*   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:234
*   | google.golang.org/grpc.(*ClientConn).Invoke
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/call.go:35
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*distSQLClient).SetupFlow
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:499
*   | github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:114
*   | github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:149
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_amd64.s:1594
* Wraps: (2) panic: use of Span after Finish. Span: setup-flow-async. Finish previously called at: <stack not captured. Set debugUseAfterFinish>
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
panic: use of Span after Finish. Span: setup-flow-async. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: setup-flow-async. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 30 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0012649f8?, {0x699cfc8, 0xc000c64840})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x4b2c1e0, 0xc005ff78d0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x99ea708?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:181 +0x12b
github.com/cockroachdb/cockroach/pkg/util/tracing.WithParent(0xc007780800)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span_options.go:204 +0x27
github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ClientInterceptor.func2({0x699cfc8, 0xc009059470}, {0x558b096, 0x2b}, {0x52a9240, 0xc00740de00}, {0x52a9380, 0xc00bbccfc0}, 0xc009059470?, 0x57f47a0, ...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:234 +0x10b
google.golang.org/grpc.(*ClientConn).Invoke(0xc008ca6ff8?, {0x699cfc8?, 0xc009059470?}, {0x558b096?, 0x0?}, {0x52a9240?, 0xc00740de00?}, {0x52a9380?, 0xc00bbccfc0?}, {0x0, ...})
	google.golang.org/grpc/external/org_golang_google_grpc/call.go:35 +0x223
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*distSQLClient).SetupFlow(0xc001264e00, {0x699cfc8, 0xc009059470}, 0xc001264e8c?, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:499 +0xc9
github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run({{0x699cfc8, 0xc009059470}, 0xc000c17580, 0xc00740de00, 0x3, 0xc005a95f20})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:114 +0x10e
github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0xc001174800, 0xc000607fa8})
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:149 +0x8f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCCLLogic_new_schema_changer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21980

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -43,6 +43,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
 	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -463,9 +464,12 @@ func (dsp *DistSQLPlanner) setupFlows(
 
 	// Start all the remote flows.
 	//
-	// usedWorker indicates whether we used at least one DistSQL worker
-	// goroutine to issue the SetupFlow RPC.
-	var usedWorker bool
+	// numAsyncRequests tracks the number of the SetupFlow RPCs that were
+	// delegated to the DistSQL runner goroutines.
+	var numAsyncRequests int
+	// numSerialRequests tracks the number of the SetupFlow RPCs that were
+	// issued by the current goroutine on its own.
+	var numSerialRequests int
 	if sp := tracing.SpanFromContext(origCtx); sp != nil && !sp.IsNoop() {
 		setupReq.TraceInfo = sp.Meta().ToProto()
 	}
@@ -490,6 +494,7 @@ func (dsp *DistSQLPlanner) setupFlows(
 	var runnerSpan *tracing.Span
 	// This span is necessary because it can outlive its parent.
 	runnerCtx, runnerSpan = tracing.ChildSpan(runnerCtx, "setup-flow-async" /* opName */)
+	// runnerCleanup can only be executed _after_ all issued RPCs are complete.
 	runnerCleanup := func() {
 		cancelRunnerCtx()
 		runnerSpan.Finish()
@@ -499,6 +504,23 @@ func (dsp *DistSQLPlanner) setupFlows(
 	var listenerGoroutineWillCleanup bool
 	defer func() {
 		if !listenerGoroutineWillCleanup {
+			// Make sure to receive from the result channel as many times as
+			// there were total SetupFlow RPCs issued, regardless of whether
+			// they were executed concurrently by a DistSQL worker or serially
+			// in the current goroutine. This is needed in order to block
+			// finishing the runner span (in runnerCleanup) until all concurrent
+			// requests are done since the runner span is used as the parent for
+			// the RPC span, and, thus, the runner span can only be finished
+			// when we know that all SetupFlow RPCs have already been completed.
+			//
+			// Note that even in case of an error in runnerRequest.run we still
+			// send on the result channel.
+			for i := 0; i < numAsyncRequests+numSerialRequests; i++ {
+				<-resultChan
+			}
+			// At this point, we know that all concurrent requests (if there
+			// were any) are complete, so we can safely perform the runner
+			// cleanup.
 			runnerCleanup()
 		}
 	}()
@@ -521,8 +543,9 @@ func (dsp *DistSQLPlanner) setupFlows(
 		// directly.
 		select {
 		case dsp.runnerCoordinator.runnerChan <- runReq:
-			usedWorker = true
+			numAsyncRequests++
 		default:
+			numSerialRequests++
 			// Use the context of the local flow since we're executing this
 			// SetupFlow RPC synchronously.
 			runReq.ctx = ctx
@@ -531,8 +554,16 @@ func (dsp *DistSQLPlanner) setupFlows(
 			}
 		}
 	}
+	if buildutil.CrdbTestBuild {
+		if numAsyncRequests+numSerialRequests != len(flows)-1 {
+			return ctx, flow, errors.AssertionFailedf(
+				"expected %d requests, found only %d async and %d serial",
+				len(flows)-1, numAsyncRequests, numSerialRequests,
+			)
+		}
+	}
 
-	if !usedWorker {
+	if numAsyncRequests == 0 {
 		// We executed all SetupFlow RPCs in the current goroutine, and all RPCs
 		// succeeded.
 		return ctx, flow, nil
@@ -565,6 +596,10 @@ func (dsp *DistSQLPlanner) setupFlows(
 		cancelRunnerCtx()
 	})
 	err = dsp.stopper.RunAsyncTask(origCtx, "distsql-remote-flows-setup-listener", func(ctx context.Context) {
+		// Note that in the loop below we always receive from the result channel
+		// as many times as there were SetupFlow RPCs issued, thus, by the time
+		// this defer is executed, we are certain that all RPCs were complete,
+		// and runnerCleanup() is safe to be executed.
 		defer runnerCleanup()
 		var seenError bool
 		for i := 0; i < len(flows)-1; i++ {

